beacon_node/operation_pool/src/lib.rs

```diff
@@ -50,6 +67,120 @@ impl<T: EthSpec> OperationPool<T> {
+    ///
+    /// ## Note
+    ///
+    /// This function assumes the given `contribution` is valid.
+    pub fn insert_sync_contribution(
+        &self,
+        contribution: SyncCommitteeContribution<T>,
+        fork: &Fork,
+        genesis_validators_root: Hash256,
+        spec: &ChainSpec,
+    ) -> Result<(), OpPoolError> {
+        let aggregate_id = SyncAggregateId::from_data::<T>(
+            contribution.slot,
+            contribution.beacon_block_root,
+            fork,
+            genesis_validators_root,
+            spec,
+        );
+
+        let mut contributions = self.sync_contributions.write();
+
+        match contributions.entry(aggregate_id) {
+            Entry::Vacant(entry) => {
+                // If no contributions or aggregate exist for these keys, insert both.
+                let contributions = vec![contribution];
+                let aggregate = SyncAggregate::from_contributions(contributions.as_slice())?;
+                entry.insert((contributions, aggregate));
+            }
+            Entry::Occupied(mut entry) => {
+                // If both a contribution and aggregate exist for this key, check whether the new or
+                // old contribution has more aggregation bits set. If the new one does, add it to the
+                // aggregate and insert both the aggregate and the contribution.
+                let existing_contributions = entry.get_mut();
+                match existing_contributions
+                    .0
+                    .iter()
+                    .position(|existing_contribution| {
+                        existing_contribution.subcommittee_index == contribution.subcommittee_index
+                    }) {
+                    Some(position) => {
+                        // Only need to recalculate if the new contribution has more bits set.
+                        if existing_contributions.0[position]
+                            .aggregation_bits
+                            .num_set_bits()
+                            < contribution.aggregation_bits.num_set_bits()
+                        {
+                            existing_contributions.0[position] = contribution;
+                            let aggregate = SyncAggregate::from_contributions(
+                                existing_contributions.0.as_slice(),
+                            )?;
+                            existing_contributions.1 = aggregate;
+                        }
+                    }
+                    None => {
+                        // If there has been no previous sync contribution for this subcommittee index,
+                        // add it and recalculate the aggregate.
+                        existing_contributions.0.push(contribution);
+                        let aggregate =
+                            SyncAggregate::from_contributions(existing_contributions.0.as_slice())?;
```

Comment:
My thinking was that we'd want to avoid calculating the aggregate right before a block proposal so we don't delay the block proposal at all. But I don't really have a sense for how long aggregate calculation will take, so may not be worth the trade-offs with recalculating it repeatedly.  I'll update this to calculate `SyncAggregate` on-demand in a separate commit from the other updates

Reply:
I just checked how long aggregations take and it's ~1.2ms for `blst` based on the _mixed addition_ benchmark from here: https://hackmd.io/@zkteam/eccbench#G2-mixed-additiondoubling2. AFAIU mixed addition is for addition of an affine point with a Jacobian one, and we have 2 Jacobians, so lets round up to 2ms per aggregation to be safe.

Given this, doing the aggregation on-demand will cost us (`SYNC_COMMITTEE_SUBNET_COUNT` - 1) * 2ms = 6ms, which seems acceptable.

On the other hand, 2ms is pretty negligible. I think the _absolute_ worst case would be if we got a better aggregate from _all_ of the target aggregators, for _all_ subcommittees. The total number of sync aggregates constructed would be `SYNC_COMMITTEE_SUBNET_COUNT * TARGET_AGGREGATORS_PER_SYNC_SUBCOMMITTEE`, and each construction would cost (`SYNC_COMMITTEE_SUBNET_COUNT` - 1) aggregations. That's `4 * 4 * 3 * 2ms` = 96ms. I think in practice it's more likely to be one or two good aggregates per subnet, which is more like `SYNC_COMMITTEE_SUBNET_COUNT * (SYNC_COMMITTEE_SUBNET_COUNT - 1) * 2ms` = 24ms. This cost is also paid by every node, rather than just the block proposer, so I'm still leaning towards the on-demand aggregation.

Would be interested to hear if @paulhauner has any thoughts.

Reply:
I had a look at what's going on here and I'd lean towards Michael's proposal; generate the `SyncAggregate` at block production time. This has the nice property that we'll only generate a single `SyncAggregate` no matter how many forks there are.

I can see you've already done this on your branch @realbigsean, so all is good as-is!